ipa-core/src/protocol/ipa_prf/quicksort.rs

```diff
@@ -54,65 +60,93 @@ where
     for<'a> &'a AdditiveShare<K>: IntoIterator<Item = AdditiveShare<K::Element>>,
     K: SharedValue + Field + CustomArray<Element = Boolean>,
 {
-    // expected amount of recursion: compute ceil(log_2(list.len()))
-    let expected_depth = std::mem::size_of::<usize>() * 8 - list.len().leading_zeros() as usize;
-    // create stack
-    let mut stack: Vec<(C, usize, usize)> = Vec::with_capacity(expected_depth);
-
-    // initialize stack
-    stack.push((ctx, 0usize, list.len()));
-
-    // iterate through quicksort recursions
-    while let Some((ctx, b_l, b_r)) = stack.pop() {
-        // start of quicksort function
-        // check whether sort is needed
-        if b_l + 1 < b_r {
-            // set up iterator
-            let mut iterator = list[b_l..b_r].iter().map(get_key);
-            // first element is pivot, apply key extraction function f
-            let pivot = iterator.next().unwrap();
-            // create pointer to context for moving into closure
-            let pctx = &(ctx.set_total_records(b_r - (b_l + 1)));
-            // precompute comparison against pivot and reveal result in parallel
-            let comp: BitVec<usize, Lsb0> = seq_join(
-                ctx.active_work(),
-                stream_iter(iterator.enumerate().map(|(n, k)| async move {
-                    // Compare the current element against pivot and reveal the result.
-                    let comparison =
-                        compare_gt(pctx.narrow(&Step::Compare), RecordId::from(n), k, pivot)
-                            .await?
-                            .reveal(pctx.narrow(&Step::Reveal), RecordId::from(n))
-                            .await?;
-
-                    // reveal outcome of comparison
-                    // desc = true will flip the order of the sort
-                    Ok::<_, Error>(Boolean::from(false ^ desc) == comparison)
-                })),
-            )
-            .try_collect()
-            .await?;
+    let mut ranges_for_next_pass = Vec::with_capacity(ranges_to_sort.len() * 2);
```

Comment:
The size of this list needs to be longer than this, or are you choosing x2 on the basis of an expected value analysis?  Or do you just not care about depth.

Reply:
Depth is not relevant here. I am only allocating a vector of ranges that will need to be sorted in the next pass. That's expected to be <= the number of ranges in the current pass.

I perform this allocation once per pass, so this will happen in expectation log(max_rows_per_user) times.